.github/workflows/windows.yaml

```diff
@@ -9,15 +9,17 @@ jobs:
       DOWNLOAD_CACHE: 'd:\downloadcache'
     runs-on: windows-latest
     steps:
+      - shell: pwsh
+        run: |
+          echo "GOPATH=D:/a/_temp/go/work" >> $env:GITHUB_ENV
+          echo "GOCACHE=D:/a/_temp/go/cache" >> $env:GITHUB_ENV
       - name: checkout
         uses: actions/checkout@v4
       - name: go-cache
         uses: actions/cache@v3
         with:
-          path: |
-            ~\AppData\Local\go-build
-            ~\go\pkg\mod
-          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          path: D:/a/_temp/go
```

Comment:
FYI - we can remove the separate actions/cache caching - that was related to https://github.com/actions/setup-go/issues/372 which is now fixed, so we should be able to use setup-go built-in caching instead of a separate step.

Reply:
The main reason I did this is, don't really like the way setup-go handles caching. I figure packing the stuff we want to cache into a single directory is a lot simpler, especially since we are special casing cache behavior just for Windows.

However, I am more interested in getting a good consensus here. What do you think out? The way I see it, we have basically three good options:
- Switch to setup-go@v4, try to get the automatic caching to work and not produce a pathological case.
- Switch to setup-go@v4 but set `cache: false` explicitly so that it's clear this is intentional.
- Keep as-is for now.

Reply:
I'm ok with keeping it as-is - it does give us more control if there are additional tweaks that could be made for actions/cache (since they may take a while to make it into actions/setup-go or it may not expose the same levels of control over caching).

Reply:
It'd be nice to try get the setup-go@v4 working with caching just to prove the pathological case still exists.

Reply:
Fair enough, will do.

Reply:
It probably won't help: https://github.com/actions/runner-images/issues/7320 😬 .

Reply:
https://github.com/bufbuild/buf/actions/runs/6907562675/job/18795193819 If you want to follow along, here's my attempt at it. Verdict: Immediately, it is clear that we are on to something here: the setup-go@v4 cache is dramatically worse. Hell, even the go test times are worse.

Reply:
Shouldn't actions/cache be removed?

Reply:
D'oh, I just realized we already do use `setup-go@v4` with `cache: false`. I must've read `cache@v4` by accident.

So I can confirm the issue is _not_ fixed using just `setup-go@v4` with caching enabled. I added a comment for why I am leaving cache disabled for now. Since `setup-go@v4` uses the `GOCACHE` and `GOMODCACHE` directories, it should be possible to make this fix work with it too, but since I'm in a stop-the-bleeding mood, I'll defer that to later unless someone would prefer I blocked on this. (More than happy to if someone feels strongly.)